plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftShutdownController.java

```diff
@@ -36,9 +42,22 @@ public OpenShiftServerBehaviour getBehavior() {
 	@Override
 	public void stop(boolean force) {
 		OpenShiftServerBehaviour behavior = getBehavior();
+		IServer server = behavior.getServer();
+		IDeploymentConfig dc = OpenShiftServerUtils.getDeploymentConfig(server);
+		boolean debugMode = OpenShiftLaunchController.DEBUG_MODE.equals(server.getMode());
+
 		behavior.setServerStopping();
+
 		try {
-			OpenShiftDebugUtils.get().terminateRemoteDebugger(behavior.getServer());
+			if(dc == null) {
+				//something is wrong, do as much as possible.
+				OpenShiftDebugUtils.get().terminateRemoteDebugger(behavior.getServer());
+			} else if(debugMode) {
+				OpenShiftLaunchController c = 
+					(OpenShiftLaunchController)behavior.getController(IControllableServerBehavior.SYSTEM_LAUNCH);
+				DebuggingContext debugContext = OpenShiftDebugUtils.get().getDebuggingContext(dc);
+				c.stopDebugging(server, dc, debugContext, new NullProgressMonitor());
```

Comment:
@fbricon I really think these should just be in utility classes.  At some point we may have multiple launch controllers depending on profile or on what we wish to do.  Expecting there to be only 1 launch controller, or expecting all launch controllers to be instances of OpenShiftLaunchController, seems a bit ambitious and hopeful.